aten/src/ATen/Allocator.h

```diff
@@ -3,14 +3,98 @@
 #include <memory>
 #include <stddef.h>
 
-#include "ATen/Retainable.h"
+#include <ATen/Error.h>
+#include <ATen/Retainable.h>
+#include <ATen/Device.h>
+#include <ATen/detail/UniqueVoidPtr.h>
 
 namespace at {
 
+// A DataPtr is a unique pointer (with an attached deleter and some
+// context for the deleter) to some memory, which also records what
+// device is for its data.
+//
+// nullptr DataPtrs can still have a nontrivial device; this allows
+// us to treat zero-size allocations uniformly with non-zero allocations.
+//
+class DataPtr {
+private:
+  detail::UniqueVoidPtr ptr_;
+  Device device_;
+public:
+  // Choice of CPU here is arbitrary; if there's an "undefined" device
+  // we could use that too
+  DataPtr() : ptr_(), device_(kCPU) {}
```

Comment:
I'm still concerned about this device -- it seems like we need to validate these are actually devices, not e.g. 'cuda'.

Reply:
It is devices; there is an implicit conversion from Backend to `at::Device`. @goldsborough, maybe it was not such a good idea to have this implicit conversion? Maybe we should make people explicitly say Device::cpu() or something like that.

Reply:
The type is called Device, that doesn't mean it's an actual device (I wanted to call it DeviceSpec).  E.g. both 'cuda:4285452' and 'cuda' are Devices but don't represent an actual device.